Add second y-axis for plotting

mattkindig opened the pull request:
### Related Issues

#11217 

### Purpose

In OMPlot, allow user to plot a variable on a second (right) y-axis

### Approach

Allows user to assign a plotted curve to the right y-axis. User can also adjust the range, and add custom axis titles to this second axis.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/OpenModelica/OpenModelica?pullRequest=14198) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/OpenModelica/OpenModelica?pullRequest=14198) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/OpenModelica/OpenModelica?pullRequest=14198) it.</sub>

adeas31 reviewed (approved):
LGTM.

Do you want to allow users to set right y-axis when using `OMPlot` directly from the command line as standalone executable. For example, see the command line arguments in main.cpp

mattkindig commented:
> Do you want to allow users to set right y-axis when using OMPlot directly from the command line as standalone executable. For example, see the command line arguments in main.cpp

Great idea! Should I convert this PR to a draft and make this change and then resubmit, or submit a separate PR once this one is merged?

mattkindig commented:
@adeas31 I have updated the commits to incorporate several new features based on the feedback in #11217, and the PR #14198 is ready for re-review. Specifically, I have implemented:

- In OMEdit, selecting the checkbox in the Variable Browser while holding down the Control (Ctrl) key causes the variable to be plotted on the right y-axis, per [this comment](https://github.com/OpenModelica/OpenModelica/issues/11217#issuecomment-3183422217)
- I added a new OMPlot command line switch `--yaxis,` which allows the user to specify one or more variables to plot on the right y-axis, per [this comment](https://github.com/OpenModelica/OpenModelica/pull/14198#pullrequestreview-3115353087).
- I also added new OMPlot command line switches `--yrange-right` and `--ylabel-right` to set the range and label for the right y-axis.

Thanks!

hkiel commented:
@adeas31 Did anyone check this with OMNotebook?
I just tried to plot a simple model, and it failed to plot signal `x`with an error promt "Invalid inputx".

adeas31 commented:
I m guessing the issue is that we added more command line options and forgot to update OMNotebook.
I will see if I can fix this.

mattkindig commented:
Sorry I'm just seeing this conversation now-- I was out of town for a few days. I admit that I hadn't tested it in OMNotebook-- in fact, I did not realize that OMNotebook and OMPlot were linked. It looks like the patch by @adeas31 fixed the issue with the command line options so that the old behavior (single y-axis) is no longer broken. Let me see if I can add in the new functionality (second y-axis support) into OMNotebook.

adeas31 commented:
Once #14268 is done you can work on changing the `PlotCallbackFunction`. Basically the plot callback is part of `threadData` that is used in the interactive API environment. Its declared in `OMCompiler/SimulationRuntime/c/gc/omc_gc.h:142`. First you need to define plot callback. We define plot callback for OMEdit in `OMEdit/OMEditLIB/OMC/OMCProxy.cpp:253` and for OMNotebook in `OMNotebook/OMNotebook/OMNotebookGUI/graphcell.cpp:1552`. Whenever, the user calls the `plot` API then we check if plot callback defined, if its defined then we use it otherwise we call OMPlot.

So start by changing the interface of plot callback in OMEdit at `OMEdit/OMEditLIB/MainWindow.h:250` and in OMNotebook at `OMNotebook/OMNotebook/OMNotebookGUI/graphcell.h:84`. Then update the code where plot callback is called from in `OMCompiler/Compiler/runtime/systemimpl.c:915`.

mattkindig commented:
@adeas31 I was loathe to change `PlotCallbackFunction` because I was worried about breaking a lot of critical code by changing an interface that is so deeply integrated into the compiler and runtime. But if you think it is a worthwhile/minimal risk activity, I can work on that. 

My thought was to define the signature as:
`static void PlotCallbackFunction(void *p, int externalWindow, const char* filename, const char* title, const char* grid, const char* plotType, const char* logX,
                                 const char* logY, const char* xLabel, const char* yLabel, const char* x1, const char* x2, const char* y1, const char* y2, const char* curveWidth,
                                 const char* curveStyle, const char* legendPosition, const char* footer, const char* autoScale, const char* additionalOptions, const char* variables);`  


Note that the only new argument is `const char* additionalOptions`, which consists of plotting options that are not already present in the existing signature (including those for the second y-axis). `additionalOptions` would be a null-terminated string consisting of `field=value`  options, separated by spaces-- that is, the same format as the command line options for OMPlot.  So for these new fields, a value of `additionalOptions` could be `--yaxis="L" --ylabel-right="RightTitle" --yrange-right=-100:200`. If there are no additional options to be specified, `NULL` is passed in for this argument when calling `PlotCallbackFunction`.

Plotting functions within OMPlot (and elsewhere) would then parse `additionalOptions` into their `field=value` pairs and apply the relevant properties to the plot.

My reasoning for doing this is that it is more extensible for the future-- the signature for `PlotCallbackFunction` stays the same, even if additional plotting options are defined down the road. Once code is updated once to this new signature, code that uses this callback within OpenModelica (or in third-party applications) only needs to be updated further if the options that being newly added are actually relevant, and any plotting options already defined in `additionalOptions` will continue to work as expected. 

This approach could avoid issues similar to the accidental breaking of OMNotebook that was observed by @hkiel -- I believe that if such an approach were to have been used, creating a simple plot in OMNotebook would have continued to work as expected (albeit without the new plotting features) rather than immediately breaking. In my opinion, we were lucky that OMNotebook did break catastrophically to detect this oversight-- changing the function signature could very well have caused silent bugs that potentially could have gone undetected for a while. My goal is to avoid this moving forward.

Before I dive into this effort, what do you think of this approach?

adeas31 commented:
Sounds good to me.

I like the approach. The only overhead with this approach is that we need to add parsing where ever plot call function is used, but I guess that is fine and intended. Perhaps we can completely re-write plot call function while we are at it i.e., replace all individual options with `const char* options`.
